WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewActivity.java

```diff
@@ -79,46 +92,122 @@ private static void openJetpackConnectionFlow(Context context, String url, SiteM
-                .getSerializableExtra(WordPress.SITE));
-        return mWebViewClient;
+        return new JetpackConnectionWebViewClient();
     }
 
     @Override
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
-        if (mWebViewClient != null) {
-            this.mWebViewClient.activityResult(this, requestCode);
+        if (requestCode == JETPACK_LOGIN && resultCode == RESULT_OK) {
+            JetpackConnectionWebViewActivity
+                    .startJetpackConnectionFlow(this, mRedirectPage, mSite, mAccountStore.hasAccessToken());
         }
+        finish();
     }
 
     @Override
     protected void cancel() {
-        if (mWebViewClient != null) {
-            this.mWebViewClient.cancel();
-        }
+        AnalyticsTracker.track(AnalyticsTracker.Stat.STATS_CANCELED_INSTALL_JETPACK);
     }
 
     @Override
     protected void onSaveInstanceState(Bundle outState) {
         super.onSaveInstanceState(outState);
-        if (mWebViewClient != null) {
-            this.mWebViewClient.onSaveInstanceState(outState);
-        }
+        outState.putString(REDIRECT_PAGE_STATE_ITEM, mRedirectPage);
     }
 
     @Override
     protected void onRestoreInstanceState(Bundle savedInstanceState) {
         super.onRestoreInstanceState(savedInstanceState);
-        if (mWebViewClient != null) {
-            this.mWebViewClient.onRestoreInstanceState(savedInstanceState);
-        }
+        mRedirectPage = savedInstanceState.getString(REDIRECT_PAGE_STATE_ITEM);
     }
 
     @Override
     public boolean onCreateOptionsMenu(Menu menu) {
         return true;
     }
+
+    private void extractRedirect(String stringUrl) throws UnsupportedEncodingException {
+        int from = stringUrl.indexOf(REDIRECT_PARAMETER) + REDIRECT_PARAMETER.length();
+        int to = stringUrl.indexOf("&", from);
+        String redirectUrl;
+        if (to > from) {
+            redirectUrl = stringUrl.substring(from, to);
+        } else {
+            redirectUrl = stringUrl.substring(from);
+        }
+        if (redirectUrl.startsWith(JETPACK_PATH)) {
+            redirectUrl = WORDPRESS_COM_PREFIX + redirectUrl;
+        }
+        mRedirectPage = URLDecoder.decode(redirectUrl, WPWebViewActivity.ENCODING_UTF8);
+    }
+
+    private void loginToJetpackInApp() {
+        Intent loginIntent = new Intent(this, LoginActivity.class);
+        LoginMode.JETPACK_STATS.putInto(loginIntent);
+        startActivityForResult(loginIntent, JETPACK_LOGIN);
+    }
+
+    private void jetpackSuccessfullyConnected(Uri uri) {
+        AnalyticsTracker.track(AnalyticsTracker.Stat.STATS_COMPLETED_INSTALL_JETPACK);
+        Intent intent = new Intent(Intent.ACTION_VIEW);
+        intent.setData(uri);
+        intent.putExtra(SITE, mSite);
+        startActivity(intent);
+        finish();
+    }
+
+    class JetpackConnectionWebViewClient extends WebViewClient {
```

Comment:
out of scope: I'd make this static and move all the logic out of it (to the activity?).